# Who will win the 2020 Masters?

This analysis focuses on predicting the 2020 Masters winner using multivariate linear regression.  The Masters is one of golf's four major championships.  Unlike American football, basketball or tennis, golf is played individually and not based on matches.  The field is make up of 96 players invited based on a set of qualifying criteria. The lowest score after four rounds of golf (72 holes) wins.  


## Data set
I chose PGA Tour ShotLink data from 1980 to 2019 (www.shotlink.com and www.pgatour.com).  This is the largest set of data available on these metrics.  The data is manually collected by volunteers who follow every PGA Tour player and record every shot made in every PGA Tour tournament.  I was looking into scraping the data when I realized someone on Kaggle had already done this (https://www.kaggle.com/bradklassen/pga-tour-20102018-data).  I instead used this data set, formatted it, and cleaned out obvious errors.

<b>Limitations:</b>
- <b>Having enough data:</b> Unfortunately, the data was only available until year end 2019.  I would have liked to have analyzed the data around player's 2020 tournament performances leading up to the Masters tournament, and create a "trending score".  Here is the explanation of why the data was delted (https://www.kaggle.com/data/174010).
- <b>Having the right variables:</b> While there are hundreds of variables tracked by the PGA tour, I would ideally like more data on how they play the golf course that hosts the Masters tournament (Augusta National Golf Club in Georgia) and potentially how they play similar holes at other courses leading up to the Masters tournament.


## Conclusion
After analyzing the lowest scorers at prior Masters tournaments and top finishers over the last 3 years, I found that P-values were high and the R-squared values were low for most variables.  There also wasn't much consistency in the results of independent variables between these two subsets of data.  This is not ideal for linear regression.  Other non-linear models (machine learning, the Elo rating system) may be better matches to predict the outcome of the 2020 Masters winner.

After trying multiple different iterations of potential models, I focused on a multivariate linear regression model that utilized the top 3 independent variables that had the lowest P-values and highest R-squared values (GIR, Scrambling and Putts).  I also used a subset of the data set that was an aggregate of the last two years of player's results at the Masters Tournament (2018 and 2019 Masters).

<b>Multivariate linear regression model:</b>  Score = 59.0038 - (0.2305 x GIR) - (0.0149 x Scrambling) + (17.4504 x Putts)

<b>Projected Winners</b> (ranked in order of who is most likely to least likely to win):
1. 	Rickie Fowler
2. 	Patrick Cantlay
3. 	Brooks Koepka
4. 	Jordan Spieth
5. 	Justin Thomas
6. 	Tiger Woods
7. 	Patton Kizzire
8. 	Dustin Johnson
9. 	Francesco Molinari
10. 	Jon Rahm
11. 	Patrick Reed
12. 	Aaron Wise
13. 	Bubba Watson
14. 	Jason Day
15. 	Tony Finau
16. 	Charley Hoffman
17. 	Louis Oosthuizen
18. 	Justin Rose
19. 	Webb Simpson
20. 	Andrew Landry

# Detailed steps shown below

## Importing, Formatting and Cleaning the Data

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
from scipy import stats
sns.set_palette("GnBu_d")

In [2]:
df = pd.read_csv('pga_data_historical.csv') #2010-2019 PGA Tour data, source: https://www.kaggle.com/bradklassen/pga-tour-20102018-data (as of August 2020, he has taken the data down)

In [3]:
df.head()

player_name        date               tournament  \
0  Rik Massengale  1980-01-13  Bob Hope Desert Classic   
1   Bobby Nichols  1980-01-13  Bob Hope Desert Classic   
2      Andy North  1980-01-13  Bob Hope Desert Classic   
3   John Mahaffey  1980-01-13  Bob Hope Desert Classic   
4  Peter Jacobsen  1980-01-13  Bob Hope Desert Classic   

                     statistic variable  value  
0  Final Round Scoring Average      AVG  70.00  
1  Final Round Scoring Average      AVG  73.00  
2  Final Round Scoring Average      AVG  73.00  
3  Final Round Scoring Average      AVG  73.00  
4  Final Round Scoring Average      AVG  73.00

In [4]:
df.shape

(46147897, 6)

In [5]:
#Unearthing the available variables in the "statistic" column
df.statistic.unique()

array(['Final Round Scoring Average', 'All-Around Ranking',
       'Par 4 Birdie or Better Leaders', 'Scoring Average Before Cut',
       'Scoring Average (Actual)', 'Round 3 Scoring Average',
       'Official Money', 'Greens in Regulation Percentage',
       'Total Birdies', 'Birdie or Better Conversion Percentage',
       'Eagles (Holes per)', 'Total Driving', 'Scoring Average',
       'Par 3 Birdie or Better Leaders', 'Driving Distance',
       'Sand Save Percentage', 'Putts Per Round',
       'Percentage of potential money won', 'Driving Accuracy Percentage',
       'Par 5 Birdie or Better Leaders', 'Par Breakers', 'Ball Striking',
       'Birdie Average', 'Total Money (Official and Unofficial)',
       'Total Eagles', 'Front 9 Round 1 Scoring Average',
       'Back 9 Round 1 Scoring Average', 'Back 9 Par 5 Scoring Average',
       'Front 9 Round 2 Scoring Average', 'Front 9 Par 4 Scoring Average',
       'Back 9 Par 3 Scoring Average', 'Front 9 Par 3 Scoring Average',
       'Fron

In [6]:
df.statistic.value_counts()

Total Eagles                               256346
Total Money (Official and Unofficial)      256156
Total Birdies                              255870
Percentage of potential money won          254714
Final Round Scoring Average                252576
                                            ...  
Late Round 5 Scoring Average                  890
Tenth Tee Early Round 5 Scoring Average       528
First Tee Late Round 5 Scoring Average        468
Tenth Tee Late Round 5 Scoring Average        422
First Tee Early Round 5 Scoring Average       276
Name: statistic, Length: 442, dtype: int64

In [7]:
#Selecting from the hundreds of variables the "statistic" column

chosen_statistic = ['Greens in Regulation Percentage',
                    'Driving Accuracy Percentage',
                    'Scrambling',
                    'Scoring Average (Actual)',
                    'Official Money',
                    'Percentage of Available Purse Won',
                    'Driving Distance',
                    'Sand Save Percentage',
                    'Overall Putting Average']

#Definitions for each variable:
    #Greens in Regulation Percentage: The percent of time a player was able to hit the green in regulation (greens hit in regulation/holes played). Note: A green is considered hit in regulation if any portion of the ball is touching the putting surface after the GIR stroke has been taken. (The GIR stroke is determined by subtracting 2 from par (1st stroke on a par 3, 2nd on a par 4, 3rd on a par 5))
    #Driving Accuracy Percentage: The percentage of time a tee shot comes to rest in the fairway (regardless of club)
    #Scrambling: The percent of time a player misses the green in regulation, but still makes par or better.
    #Scoring Average (Actual): The average number of strokes per completed round.
    #Official Money: The total official money a player has earned year-to-date. Note: This is for PGA TOUR members only.
    #Percentage of Available Purse Won: For official events, the player's total money won as a percentage of the total purse available.
    #Driving Distance: The average number of yards per measured drive. These drives are measured on two holes per round. Care is taken to select two holes which face in opposite directions to counteract the effect of wind. Drives are measured to the point at which they come to rest regardless of whether they are in the fairway or not.
    #Sand Save Percentage: The percent of time a player was able to get 'up and down' once in a greenside sand bunker (regardless of score). Note: 'Up and down' indicates it took the player 2 shots or less to put the ball in the hole from that point.
    #Overall Putting Average: The average number of putts for all holes played (total putts / total holes played).

In [8]:
#Adjusting the DataFrame
df = df[df['statistic'].isin(chosen_statistic)]
df = df[df['variable'] != 'RANK THIS WEEK'] #Multiple variables have the value 'RANK THIS WEEK', so unstacking will create duplicate columns of 'RANK THIS WEEK', causing an error ("Index contains duplicate entries, cannot reshape"). This line eliminates the error
df = df.drop(columns = ['variable'])
df.head()

player_name        date               tournament  \
145     Bill Rogers  1980-01-13  Bob Hope Desert Classic   
146      Gil Morgan  1980-01-13  Bob Hope Desert Classic   
147      Bob Gilder  1980-01-13  Bob Hope Desert Classic   
148  Billy Kratzert  1980-01-13  Bob Hope Desert Classic   
149   Roger Maltbie  1980-01-13  Bob Hope Desert Classic   

                    statistic  value  
145  Scoring Average (Actual)  70.60  
146  Scoring Average (Actual)  70.60  
147  Scoring Average (Actual)  70.60  
148  Scoring Average (Actual)  70.60  
149  Scoring Average (Actual)  70.60

In [9]:
#Unstacking the data in the DataFrame
df = df.set_index(['player_name', 'date', 'tournament','statistic'])['value'].unstack('statistic').reset_index()
df.head()

statistic     player_name        date                            tournament  \
0             A.J. Duncan  1988-04-03         KMart Greater Greensboro Open   
1             A.J. Duncan  1990-03-11                         Honda Classic   
2          A.J. McInerney  2017-11-05  Shriners Hospitals for Children Open   
3          A.J. McInerney  2018-06-10                FedEx St. Jude Classic   
4          Aaron Baddeley  2000-03-12                         Honda Classic   

statistic Driving Accuracy Percentage Driving Distance  \
0                               58.93            283.5   
1                               44.64            264.1   
2                               50.00            317.8   
3                               50.00            298.5   
4                               64.29            283.6   

statistic Greens in Regulation Percentage Official Money  \
0                                   58.33         $1,810   
1                                   52.78         $1,810   
2                                   76.39            NaN   
3                                   56.94            NaN   
4                                   70.83            NaN   

statistic Overall Putting Average Percentage of Available Purse Won  \
0                             NaN                               .18   
1                             NaN                               .18   
2                           1.694                              2.22   
3                           1.514                               .22   
4                           1.625                               NaN   

statistic Sand Save Percentage Scoring Average (Actual) Scrambling  
0                        33.33                    73.75        NaN  
1                        14.29                    77.00        NaN  
2                        22.22                    69.50      52.94  
3                        33.33                    70.75      61.29  
4                        50.00                    70.25      66.67

In [10]:
df.shape

(128160, 12)

In [11]:
df.describe()

statistic     player_name        date        tournament  \
count              128160      128160            128160   
unique               2440        1697               304   
top        Davis Love III  1995-07-23  PGA Championship   
freq                  546         178              3018   

statistic Driving Accuracy Percentage Driving Distance  \
count                          124669           123306   
unique                            333             1169   
top                             64.29            271.9   
freq                             7030              322   

statistic Greens in Regulation Percentage Official Money  \
count                              124543         123674   
unique                                111          11277   
top                                 66.67              $   
freq                                 8056            200   

statistic Overall Putting Average Percentage of Available Purse Won  \
count                       87568                            118148   
unique                        206                               772   
top                         1.611                               .22   
freq                         6397                              7025   

statistic Sand Save Percentage Scoring Average (Actual) Scrambling  
count                   123011                   126125      87574  
unique                     134                      129        425  
top                      50.00                    70.00      66.67  
freq                     19184                     6630       4253

## Fixing the Data types

In [12]:
print(df.dtypes)

statistic
player_name                          object
date                                 object
tournament                           object
Driving Accuracy Percentage          object
Driving Distance                     object
Greens in Regulation Percentage      object
Official Money                       object
Overall Putting Average              object
Percentage of Available Purse Won    object
Sand Save Percentage                 object
Scoring Average (Actual)             object
Scrambling                           object
dtype: object


In [13]:
df.head()

statistic     player_name        date                            tournament  \
0             A.J. Duncan  1988-04-03         KMart Greater Greensboro Open   
1             A.J. Duncan  1990-03-11                         Honda Classic   
2          A.J. McInerney  2017-11-05  Shriners Hospitals for Children Open   
3          A.J. McInerney  2018-06-10                FedEx St. Jude Classic   
4          Aaron Baddeley  2000-03-12                         Honda Classic   

statistic Driving Accuracy Percentage Driving Distance  \
0                               58.93            283.5   
1                               44.64            264.1   
2                               50.00            317.8   
3                               50.00            298.5   
4                               64.29            283.6   

statistic Greens in Regulation Percentage Official Money  \
0                                   58.33         $1,810   
1                                   52.78         $1,810   
2                                   76.39            NaN   
3                                   56.94            NaN   
4                                   70.83            NaN   

statistic Overall Putting Average Percentage of Available Purse Won  \
0                             NaN                               .18   
1                             NaN                               .18   
2                           1.694                              2.22   
3                           1.514                               .22   
4                           1.625                               NaN   

statistic Sand Save Percentage Scoring Average (Actual) Scrambling  
0                        33.33                    73.75        NaN  
1                        14.29                    77.00        NaN  
2                        22.22                    69.50      52.94  
3                        33.33                    70.75      61.29  
4                        50.00                    70.25      66.67

In [14]:
#Cleaning the "Official Money" column so we can analyze it
df['Official Money'] = df['Official Money'].str.replace('\$', '', regex = True)
df['Official Money'] = df['Official Money'].str.replace(',', '', regex = True)
df['Official Money'] = df['Official Money'].str.replace('O', '0', regex = True)
df['Official Money'] = df['Official Money'].str.replace(' ', '', regex = True)
df['Official Money'] = df['Official Money'].str.strip()
df['Official Money'].replace("", np.nan, inplace=True)

In [15]:
#Converting data types from objects to float
df['Greens in Regulation Percentage']=df['Greens in Regulation Percentage'].astype('float')
df['Driving Accuracy Percentage']=df['Driving Accuracy Percentage'].astype('float')
df['Scrambling']=df['Scrambling'].astype('float')
df['Scoring Average (Actual)']=df['Scoring Average (Actual)'].astype('float')
df['Official Money']=df['Official Money'].astype('float')
df['Percentage of Available Purse Won']=df['Percentage of Available Purse Won'].astype('float')
df['Driving Distance']=df['Driving Distance'].astype('float')
df['Sand Save Percentage']=df['Sand Save Percentage'].astype('float')
df['Overall Putting Average']=df['Overall Putting Average'].astype('float')

In [16]:
print(df.dtypes)

statistic
player_name                           object
date                                  object
tournament                            object
Driving Accuracy Percentage          float64
Driving Distance                     float64
Greens in Regulation Percentage      float64
Official Money                       float64
Overall Putting Average              float64
Percentage of Available Purse Won    float64
Sand Save Percentage                 float64
Scoring Average (Actual)             float64
Scrambling                           float64
dtype: object


In [17]:
#Cleaning up the DataFrame - renaming certain columns
df.rename(columns = {'Greens in Regulation Percentage':'GIR',
                     'Driving Accuracy Percentage':'Fairways',
                     'Scrambling':'Scrambling',
                     'Scoring Average (Actual)':'Score',
                     'Official Money':'Money',
                     'Percentage of Available Purse Won':'Purse',
                     'Driving Distance':'Distance',
                     'Sand Save Percentage':'Sandies',
                     'Overall Putting Average':'Putts'}, inplace=True)

In [18]:
df.head()

statistic     player_name        date                            tournament  \
0             A.J. Duncan  1988-04-03         KMart Greater Greensboro Open   
1             A.J. Duncan  1990-03-11                         Honda Classic   
2          A.J. McInerney  2017-11-05  Shriners Hospitals for Children Open   
3          A.J. McInerney  2018-06-10                FedEx St. Jude Classic   
4          Aaron Baddeley  2000-03-12                         Honda Classic   

statistic  Fairways  Distance    GIR   Money  Putts  Purse  Sandies  Score  \
0             58.93     283.5  58.33  1810.0    NaN   0.18    33.33  73.75   
1             44.64     264.1  52.78  1810.0    NaN   0.18    14.29  77.00   
2             50.00     317.8  76.39     NaN  1.694   2.22    22.22  69.50   
3             50.00     298.5  56.94     NaN  1.514   0.22    33.33  70.75   
4             64.29     283.6  70.83     NaN  1.625    NaN    50.00  70.25   

statistic  Scrambling  
0                 NaN  
1                 NaN  
2               52.94  
3               61.29  
4               66.67

In [19]:
df.describe()

statistic       Fairways       Distance            GIR         Money  \
count      124669.000000  123306.000000  124543.000000  1.234740e+05   
mean           65.607781     275.949496      66.555930  4.870611e+04   
std            10.898736      19.243917       8.459017  1.231114e+05   
min            17.860000     182.600000      20.830000  9.600000e+01   
25%            58.930000     261.500000      61.110000  4.265000e+03   
50%            66.070000     275.300000      66.670000  1.332500e+04   
75%            73.210000     289.800000      72.220000  3.987500e+04   
max           100.000000     362.100000     487.500000  2.250000e+06   

statistic         Putts          Purse        Sandies          Score  \
count      87568.000000  118148.000000  123011.000000  126125.000000   
mean           1.604443       1.407430      51.691787      70.651087   
std            0.084746       2.585814      24.156935       1.862798   
min            0.611000       0.130000       0.000000      63.250000   
25%            1.556000       0.230000      36.360000      69.250000   
50%            1.611000       0.520000      50.000000      70.500000   
75%            1.653000       1.450000      66.670000      71.750000   
max            2.056000      30.000000     200.000000      81.250000   

statistic    Scrambling  
count      87574.000000  
mean          61.097847  
std           11.346915  
min            4.170000  
25%           53.570000  
50%           61.290000  
75%           68.750000  
max          100.000000

### Fixing errors in the data
There is at least one GIR value that is above 100% (in the chart above the max = 487.5) and one Sandies value that is above 100% (in the chart above the max = 200.0).  Will drop any rows where GIR or Sandies is above 100%:

In [20]:
df=df[df.GIR <= 100]
df=df[df.Sandies <= 100]

In [21]:
df.describe()

statistic       Fairways       Distance            GIR         Money  \
count      122818.000000  121486.000000  122848.000000  1.184600e+05   
mean           65.584648     276.036268      66.498876  4.843005e+04   
std            10.904967      19.236936       8.347419  1.213998e+05   
min            17.860000     182.600000      20.830000  2.800000e+02   
25%            58.930000     261.500000      61.110000  4.332000e+03   
50%            66.070000     275.400000      66.670000  1.339200e+04   
75%            73.210000     289.900000      72.220000  3.990000e+04   
max           100.000000     362.100000     100.000000  2.250000e+06   

statistic         Putts          Purse        Sandies          Score  \
count      86385.000000  113663.000000  122848.000000  122691.000000   
mean           1.604132       1.391657      51.688042      70.651976   
std            0.084406       2.558169      24.150509       1.852941   
min            0.611000       0.130000       0.000000      63.250000   
25%            1.556000       0.230000      36.360000      69.330000   
50%            1.611000       0.520000      50.000000      70.500000   
75%            1.653000       1.450000      66.670000      71.750000   
max            2.056000      30.000000     100.000000      81.250000   

statistic    Scrambling  
count      86391.000000  
mean          61.093638  
std           11.312598  
min            4.170000  
25%           53.570000  
50%           61.290000  
75%           68.750000  
max          100.000000

In [22]:
#Adding the year as a column to the dataframe
df['year'] = pd.DatetimeIndex(df['date']).year
df['month'] = pd.DatetimeIndex(df['date']).month
df.head()

statistic     player_name        date                            tournament  \
0             A.J. Duncan  1988-04-03         KMart Greater Greensboro Open   
1             A.J. Duncan  1990-03-11                         Honda Classic   
2          A.J. McInerney  2017-11-05  Shriners Hospitals for Children Open   
3          A.J. McInerney  2018-06-10                FedEx St. Jude Classic   
4          Aaron Baddeley  2000-03-12                         Honda Classic   

statistic  Fairways  Distance    GIR   Money  Putts  Purse  Sandies  Score  \
0             58.93     283.5  58.33  1810.0    NaN   0.18    33.33  73.75   
1             44.64     264.1  52.78  1810.0    NaN   0.18    14.29  77.00   
2             50.00     317.8  76.39     NaN  1.694   2.22    22.22  69.50   
3             50.00     298.5  56.94     NaN  1.514   0.22    33.33  70.75   
4             64.29     283.6  70.83     NaN  1.625    NaN    50.00  70.25   

statistic  Scrambling  year  month  
0                 NaN  1988      4  
1                 NaN  1990      3  
2               52.94  2017     11  
3               61.29  2018      6  
4               66.67  2000      3

# Predicting the 2020 Masters winner

## Reviewing who scored the lowest in Masters Tournaments over the last decade
For simplicity, this reviews any players who scored 69 or less on average for each round of the tournament.  Par is 72 at the golf course that hosts the Masters Tournament (Augusta National), so these players scored 3 strokes less than par on average.

In [23]:
df_masters_winners = df.loc[(df['tournament'] == 'Masters Tournament') & (df.year > 2009) & (df.Score <= 69)]
df_masters_winners.nsmallest(50, 'Score')

statistic        player_name        date          tournament  Fairways  \
68943          Jordan Spieth  2015-04-12  Masters Tournament     69.64   
98008         Phil Mickelson  2010-04-11  Masters Tournament     60.71   
94192           Patrick Reed  2018-04-08  Masters Tournament     73.21   
70128            Justin Rose  2015-04-12  Masters Tournament     78.57   
100693         Rickie Fowler  2018-04-08  Masters Tournament     71.43   
21310       Charl Schwartzel  2011-04-10  Masters Tournament     66.07   
98091         Phil Mickelson  2015-04-12  Masters Tournament     69.64   
69000          Jordan Spieth  2018-04-08  Masters Tournament     67.86   
117798           Tiger Woods  2019-04-14  Masters Tournament     62.50   
77684           Lee Westwood  2010-04-11  Masters Tournament     62.50   
104514          Rory McIlroy  2015-04-12  Masters Tournament     76.79   
652               Adam Scott  2011-04-10  Masters Tournament     62.50   
38616         Dustin Johnson  2019-04-14  Masters Tournament     60.71   
17525          Brooks Koepka  2019-04-14  Masters Tournament     69.64   
2917             Anthony Kim  2010-04-11  Masters Tournament     58.93   
54685              Jason Day  2011-04-10  Masters Tournament     69.64   
127350     Xander Schauffele  2019-04-14  Masters Tournament     62.50   

statistic  Distance    GIR      Money  Putts  Purse  Sandies  Score  \
68943         282.6  75.00  1800000.0  1.500  18.00    50.00  67.50   
98008         297.1  75.00  1350000.0  1.611  18.00     0.00  68.00   
94192         299.3  66.67  1980000.0  1.444  18.00    50.00  68.25   
70128         293.8  75.00   880000.0  1.611   8.80    66.67  68.50   
100693        290.6  70.83  1188000.0  1.569  10.80    50.00  68.50   
21310         278.4  68.06  1440000.0  1.486  19.20    42.86  68.50   
98091         293.9  70.83   880000.0  1.583   8.80    57.14  68.50   
69000         287.1  72.22   748000.0  1.625   6.80    66.67  68.75   
117798        294.6  80.56  2070000.0  1.667    NaN    66.67  68.75   
77684         292.3  80.56        NaN  1.681  10.80    60.00  68.75   
104514        297.0  73.61   480000.0  1.639   4.80    66.67  69.00   
652           287.1  70.83   704000.0  1.542   9.39    60.00  69.00   
38616         308.0  70.83   858667.0  1.569    NaN   100.00  69.00   
17525         313.6  73.61   858667.0  1.639    NaN   100.00  69.00   
2917          277.3  70.83   510000.0  1.583   6.80    75.00  69.00   
54685         287.3  72.22   704000.0  1.583   9.39   100.00  69.00   
127350        305.8  70.83   858667.0  1.597    NaN    57.14  69.00   

statistic  Scrambling  year  month  
68943           66.67  2015      4  
98008           77.78  2010      4  
94192           62.50  2018      4  
70128           61.11  2015      4  
100693          76.19  2018      4  
21310           78.26  2011      4  
98091           76.19  2015      4  
69000           70.00  2018      4  
117798          50.00  2019      4  
77684           64.29  2010      4  
104514          57.89  2015      4  
652             61.90  2011      4  
38616           80.95  2019      4  
17525           68.42  2019      4  
2917            61.90  2010      4  
54685           60.00  2011      4  
127350          57.14  2019      4

In [24]:
#Fairways
Winners_Fairways_model = sm.OLS.from_formula("Score ~ Fairways", data=df_masters_winners)
Winners_Fairways_results = Winners_Fairways_model.fit()
print("Score + Fairways | P-value:", Winners_Fairways_results.pvalues[1]/2, "| R-squared:", Winners_Fairways_results.rsquared)

#Distance
Winners_Distance_model = sm.OLS.from_formula("Score ~ Distance", data=df_masters_winners)
Winners_Distance_results = Winners_Distance_model.fit()
print("Score + Distance | P-value:", Winners_Distance_results.pvalues[1]/2, "| R-squared:", Winners_Distance_results.rsquared)

#GIR
Winners_GIR_model = sm.OLS.from_formula("Score ~ GIR", data=df_masters_winners)
Winners_GIR_results = Winners_GIR_model.fit()
print("Score + GIR | P-value:", Winners_GIR_results.pvalues[1]/2, "| R-squared:", Winners_GIR_results.rsquared)

#Putts
Winners_Putts_model = sm.OLS.from_formula("Score ~ Putts", data=df_masters_winners)
Winners_Putts_results = Winners_Putts_model.fit()
print("Score + Putts | P-value:", Winners_Putts_results.pvalues[1]/2, "| R-squared:", Winners_Putts_results.rsquared)

#Sandies
Winners_Sandies_model = sm.OLS.from_formula("Score ~ Sandies", data=df_masters_winners)
Winners_Sandies_results = Winners_Sandies_model.fit()
print("Score + Sandies | P-value:", Winners_Sandies_results.pvalues[1]/2, "| R-squared:", Winners_Sandies_results.rsquared)

#Scrambling
Winners_Scrambling_model = sm.OLS.from_formula("Score ~ Scrambling", data=df_masters_winners)
Winners_Scrambling_results = Winners_Scrambling_model.fit()
print("Score + Scrambling | P-value:", Winners_Scrambling_results.pvalues[1]/2, "| R-squared:", Winners_Scrambling_results.rsquared)

#Money
Winners_Money_model = sm.OLS.from_formula("Score ~ Money", data=df_masters_winners)
Winners_Money_results = Winners_Money_model.fit()
print("Score + Money | P-value:", Winners_Money_results.pvalues[1]/2, "| R-squared:", Winners_Money_results.rsquared)


#Note:
#P-value is the probability of getting a sample proportion at least this extreme.  Ranges from zero to 100%. 
#The P>|t| value in the OLS summary reflects a two-sided P-value, and needs to be divided by 2 to reflect the single sided value.
#P-values quoted in the summary of this notebook reflect the single sided P-value


Score + Fairways | P-value: 0.2240475033709749 | R-squared: 0.038883150137074884
Score + Distance | P-value: 0.15807395452360568 | R-squared: 0.06689997097484623
Score + GIR | P-value: 0.41398534430943346 | R-squared: 0.0032494005209643406
Score + Putts | P-value: 0.035671100457042304 | R-squared: 0.2006769600829218
Score + Sandies | P-value: 0.0019330609663386158 | R-squared: 0.4368900571096984
Score + Scrambling | P-value: 0.12970249401895406 | R-squared: 0.08391424355080668
Score + Money | P-value: 0.002362477426547019 | R-squared: 0.44554907438337543


In [25]:
df_masters_winners.corr()

statistic   Fairways  Distance       GIR     Money     Putts     Purse  \
statistic                                                                
Fairways    1.000000  0.076079 -0.132931 -0.009744 -0.133614 -0.157185   
Distance    0.076079  1.000000  0.073070 -0.036656  0.297111 -0.036458   
GIR        -0.132931  0.073070  1.000000  0.199351  0.755508 -0.190875   
Money      -0.009744 -0.036656  0.199351  1.000000 -0.396587  0.913606   
Putts      -0.133614  0.297111  0.755508 -0.396587  1.000000 -0.663060   
Purse      -0.157185 -0.036458 -0.190875  0.913606 -0.663060  1.000000   
Sandies     0.094395  0.257322 -0.007945 -0.428993  0.218628 -0.657227   
Score      -0.197188  0.258650 -0.057004 -0.667495  0.447970 -0.766933   
Scrambling -0.122191  0.015398 -0.375639 -0.045039 -0.284665  0.466345   
year        0.263912  0.624001 -0.084440  0.238100  0.040814 -0.082705   
month            NaN       NaN       NaN       NaN       NaN       NaN   

statistic    Sandies     Score  Scrambling      year  month  
statistic                                                    
Fairways    0.094395 -0.197188   -0.122191  0.263912    NaN  
Distance    0.257322  0.258650    0.015398  0.624001    NaN  
GIR        -0.007945 -0.057004   -0.375639 -0.084440    NaN  
Money      -0.428993 -0.667495   -0.045039  0.238100    NaN  
Putts       0.218628  0.447970   -0.284665  0.040814    NaN  
Purse      -0.657227 -0.766933    0.466345 -0.082705    NaN  
Sandies     1.000000  0.660977   -0.233590  0.308651    NaN  
Score       0.660977  1.000000   -0.289680  0.082350    NaN  
Scrambling -0.233590 -0.289680    1.000000 -0.068381    NaN  
year        0.308651  0.082350   -0.068381  1.000000    NaN  
month            NaN       NaN         NaN       NaN    NaN

### Summary
- Nearly all the P-values are above alpha of 0.05 except for Sandies (0.002) and Putts (0.036).
- R-squared for Sandies (0.44) and Putts (0.20) are the highest for the variables available.
- The average number of putts were 1.681 or lower
- With the exception of Mickelson in 2010 and Schwartzel in 2011, sand saves were above 50%.  Sand saves are difficult to use as restriction to an analysis, because they are only relevant if the player misses the green (opposite of a GIR) and the ball lands in the sand.  In addition, it is difficult to apply a player's prior Sandies record to their future 2020 Master's performance outside of prior Masters tournaments because the sand used in the bunkers are not what is typically used at other golf courses (it is made of material from feldspar mines in North Carolina).
- Note that the summary analysis excludes Money, because the Money represents how much money they won at the Masters Tournament

## Analyzing prior Masters Tournament top finishers by year

### 2017

In [26]:
#2017
#Sergio Garcia won in the playoff
df_2017_winners = df.loc[(df['tournament'] == 'Masters Tournament') & (df['year']==2017)]
df_2017_winners = df_2017_winners.nsmallest(10, 'Score')
df_2017_winners.head(10)

statistic       player_name        date          tournament  Fairways  \
70164           Justin Rose  2017-04-09  Masters Tournament     62.50   
110292        Sergio Garcia  2017-04-09  Masters Tournament     80.36   
21401      Charl Schwartzel  2017-04-09  Masters Tournament     69.64   
85495           Matt Kuchar  2017-04-09  Masters Tournament     67.86   
117410       Thomas Pieters  2017-04-09  Masters Tournament     60.71   
95037            Paul Casey  2017-04-09  Masters Tournament     62.50   
73210        Kevin Chappell  2017-04-09  Masters Tournament     60.71   
104543         Rory McIlroy  2017-04-09  Masters Tournament     51.79   
744              Adam Scott  2017-04-09  Masters Tournament     64.29   
106010           Ryan Moore  2017-04-09  Masters Tournament     75.00   

statistic  Distance    GIR      Money  Putts  Purse  Sandies  Score  \
70164         279.8  75.00  1188000.0  1.667  11.88     0.00  69.75   
110292        291.9  75.00  1980000.0  1.653  19.80    83.33  69.75   
21401         280.8  68.06   748000.0  1.625   7.48    60.00  70.50   
85495         269.9  62.50   484000.0  1.556   4.84    33.33  70.75   
117410        293.0  66.67        NaN  1.611   4.84    44.44  70.75   
95037         277.9  77.78   396000.0  1.764   3.96     0.00  71.00   
73210         288.5  70.83   354750.0  1.708   3.55    20.00  71.25   
104543        288.0  61.11   354750.0  1.611   3.55    42.86  71.25   
744           291.0  73.61   308000.0  1.722   3.08    57.14  71.50   
106010        276.8  63.89   308000.0  1.653   3.08    50.00  71.50   

statistic  Scrambling  year  month  
70164           50.00  2017      4  
110292          66.67  2017      4  
21401           56.52  2017      4  
85495           70.37  2017      4  
117410          58.33  2017      4  
95037           56.25  2017      4  
73210           42.86  2017      4  
104543          75.00  2017      4  
744             47.37  2017      4  
106010          61.54  2017      4

In [27]:
#Fairways
winners17_Fairways_model = sm.OLS.from_formula("Score ~ Fairways", data=df_2017_winners)
winners17_Fairways_results = winners17_Fairways_model.fit()
print("Score + Fairways | P-value:", winners17_Fairways_results.pvalues[1]/2, "| R-squared:", winners17_Fairways_results.rsquared)

#Distance
winners17_Distance_model = sm.OLS.from_formula("Score ~ Distance", data=df_2017_winners)
winners17_Distance_results = winners17_Distance_model.fit()
print("Score + Distance | P-value:", winners17_Distance_results.pvalues[1]/2, "| R-squared:", winners17_Distance_results.rsquared)

#GIR
winners17_GIR_model = sm.OLS.from_formula("Score ~ GIR", data=df_2017_winners)
winners17_GIR_results = winners17_GIR_model.fit()
print("Score + GIR | P-value:", winners17_GIR_results.pvalues[1]/2, "| R-squared:", winners17_GIR_results.rsquared)

#Putts
winners17_Putts_model = sm.OLS.from_formula("Score ~ Putts", data=df_2017_winners)
winners17_Putts_results = winners17_Putts_model.fit()
print("Score + Putts | P-value:", winners17_Putts_results.pvalues[1]/2, "| R-squared:", winners17_Putts_results.rsquared)

#Sandies
winners17_Sandies_model = sm.OLS.from_formula("Score ~ Sandies", data=df_2017_winners)
winners17_Sandies_results = winners17_Sandies_model.fit()
print("Score + Sandies | P-value:", winners17_Sandies_results.pvalues[1]/2, "| R-squared:", winners17_Sandies_results.rsquared)

#Scrambling
winners17_Scrambling_model = sm.OLS.from_formula("Score ~ Scrambling", data=df_2017_winners)
winners17_Scrambling_results = winners17_Scrambling_model.fit()
print("Score + Scrambling | P-value:", winners17_Scrambling_results.pvalues[1]/2, "| R-squared:", winners17_Scrambling_results.rsquared)

#Money
winners17_Money_model = sm.OLS.from_formula("Score ~ Money", data=df_2017_winners)
winners17_Money_results = winners17_Money_model.fit()
print("Score + Money | P-value:", winners17_Money_results.pvalues[1]/2, "| R-squared:", winners17_Money_results.rsquared)


#Note:
#P-value is the probability of getting a sample proportion at least this extreme.  Ranges from zero to 100%. 
#The P>|t| value in the OLS summary reflects a two-sided P-value, and needs to be divided by 2 to reflect the single sided value.
#P-values quoted in the summary of this notebook reflect the single sided P-value

Score + Fairways | P-value: 0.15559497919745446 | R-squared: 0.1274599195858377
Score + Distance | P-value: 0.49501838002690585 | R-squared: 2.0745460284632422e-05
Score + GIR | P-value: 0.13312562098999206 | R-squared: 0.1514993235012131
Score + Putts | P-value: 0.27370999525603185 | R-squared: 0.04700206273177776
Score + Sandies | P-value: 0.45663874723993125 | R-squared: 0.001576655314487141
Score + Scrambling | P-value: 0.4014889669113523 | R-squared: 0.008246901643381377
Score + Money | P-value: 0.00046596396807391336 | R-squared: 0.8106279188767553


### 2018

In [28]:
#2018 
df_2018_winners = df.loc[(df['tournament'] == 'Masters Tournament') & (df['year']==2018)]
df_2018_winners = df_2018_winners.nsmallest(10, 'Score')
df_2018_winners.head(10)

statistic     player_name        date          tournament  Fairways  Distance  \
94192        Patrick Reed  2018-04-08  Masters Tournament     73.21     299.3   
100693      Rickie Fowler  2018-04-08  Masters Tournament     71.43     290.6   
69000       Jordan Spieth  2018-04-08  Masters Tournament     67.86     287.1   
68287            Jon Rahm  2018-04-08  Masters Tournament     73.21     295.4   
18650        Bubba Watson  2018-04-08  Masters Tournament     83.93     304.0   
19797       Cameron Smith  2018-04-08  Masters Tournament     58.93     293.3   
50235      Henrik Stenson  2018-04-08  Masters Tournament     76.79     284.9   
104555       Rory McIlroy  2018-04-08  Masters Tournament     62.50     302.6   
80513       Marc Leishman  2018-04-08  Masters Tournament     53.57     291.8   
38593      Dustin Johnson  2018-04-08  Masters Tournament     66.07     304.9   

statistic    GIR      Money  Putts  Purse  Sandies  Score  Scrambling  year  \
94192      66.67  1980000.0  1.444  18.00    50.00  68.25       62.50  2018   
100693     70.83  1188000.0  1.569  10.80    50.00  68.50       76.19  2018   
69000      72.22   748000.0  1.625   6.80    66.67  68.75       70.00  2018   
68287      68.06   528000.0  1.556   4.80    75.00  69.25       60.87  2018   
18650      77.78   386375.0  1.694   3.51    60.00  69.75       62.50  2018   
19797      68.06   386375.0  1.583   3.51    40.00  69.75       69.57  2018   
50235      70.83   386375.0  1.625   3.51    50.00  69.75       57.14  2018   
104555     59.72   386375.0  1.514   3.51    80.00  69.75       75.86  2018   
80513      65.28   319000.0  1.569   2.90    66.67  70.00       72.00  2018   
38593      68.06   286000.0  1.667   2.60    50.00  70.25       65.22  2018   

statistic  month  
94192          4  
100693         4  
69000          4  
68287          4  
18650          4  
19797          4  
50235          4  
104555         4  
80513          4  
38593          4

In [29]:
#Fairways
winners18_Fairways_model = sm.OLS.from_formula("Score ~ Fairways", data=df_2018_winners)
winners18_Fairways_results = winners18_Fairways_model.fit()
print("Score + Fairways | P-value:", winners18_Fairways_results.pvalues[1]/2, "| R-squared:", winners18_Fairways_results.rsquared)

#Distance
winners18_Distance_model = sm.OLS.from_formula("Score ~ Distance", data=df_2018_winners)
winners18_Distance_results = winners18_Distance_model.fit()
print("Score + Distance | P-value:", winners18_Distance_results.pvalues[1]/2, "| R-squared:", winners18_Distance_results.rsquared)

#GIR
winners18_GIR_model = sm.OLS.from_formula("Score ~ GIR", data=df_2018_winners)
winners18_GIR_results = winners18_GIR_model.fit()
print("Score + GIR | P-value:", winners18_GIR_results.pvalues[1]/2, "| R-squared:", winners18_GIR_results.rsquared)

#Putts
winners18_Putts_model = sm.OLS.from_formula("Score ~ Putts", data=df_2018_winners)
winners18_Putts_results = winners18_Putts_model.fit()
print("Score + Putts | P-value:", winners18_Putts_results.pvalues[1]/2, "| R-squared:", winners18_Putts_results.rsquared)

#Sandies
winners18_Sandies_model = sm.OLS.from_formula("Score ~ Sandies", data=df_2018_winners)
winners18_Sandies_results = winners18_Sandies_model.fit()
print("Score + Sandies | P-value:", winners18_Sandies_results.pvalues[1]/2, "| R-squared:", winners18_Sandies_results.rsquared)

#Scrambling
winners18_Scrambling_model = sm.OLS.from_formula("Score ~ Scrambling", data=df_2018_winners)
winners18_Scrambling_results = winners18_Scrambling_model.fit()
print("Score + Scrambling | P-value:", winners18_Scrambling_results.pvalues[1]/2, "| R-squared:", winners18_Scrambling_results.rsquared)

#Money
winners18_Money_model = sm.OLS.from_formula("Score ~ Money", data=df_2018_winners)
winners18_Money_results = winners18_Money_model.fit()
print("Score + Money | P-value:", winners18_Money_results.pvalues[1]/2, "| R-squared:", winners18_Money_results.rsquared)


#Note:
#P-value is the probability of getting a sample proportion at least this extreme.  Ranges from zero to 100%. 
#The P>|t| value in the OLS summary reflects a two-sided P-value, and needs to be divided by 2 to reflect the single sided value.
#P-values quoted in the summary of this notebook reflect the single sided P-value

Score + Fairways | P-value: 0.19739577117732404 | R-squared: 0.09179765405919837
Score + Distance | P-value: 0.21361245814094232 | R-squared: 0.08040710586512434
Score + GIR | P-value: 0.3582503900023769 | R-squared: 0.01738941757680501
Score + Putts | P-value: 0.05322932035005617 | R-squared: 0.29252273556899966
Score + Sandies | P-value: 0.41845116329306054 | R-squared: 0.005621856116004853
Score + Scrambling | P-value: 0.4001296028615949 | R-squared: 0.008480017718602406
Score + Money | P-value: 0.00024682507331018165 | R-squared: 0.7983813756913664


### 2019

In [30]:
#2019
df_2019_winners = df.loc[(df['tournament'] == 'Masters Tournament') & (df['year']==2019)]
df_2019_winners = df_2019_winners.nsmallest(10, 'Score')
df_2019_winners.head(10)

statistic         player_name        date          tournament  Fairways  \
117798            Tiger Woods  2019-04-14  Masters Tournament     62.50   
17525           Brooks Koepka  2019-04-14  Masters Tournament     69.64   
38616          Dustin Johnson  2019-04-14  Masters Tournament     60.71   
127350      Xander Schauffele  2019-04-14  Masters Tournament     62.50   
40810      Francesco Molinari  2019-04-14  Masters Tournament     73.21   
54817               Jason Day  2019-04-14  Masters Tournament     66.07   
123063             Tony Finau  2019-04-14  Masters Tournament     67.86   
126086           Webb Simpson  2019-04-14  Masters Tournament     83.93   
68306                Jon Rahm  2019-04-14  Masters Tournament     76.79   
94045         Patrick Cantlay  2019-04-14  Masters Tournament     64.29   

statistic  Distance    GIR      Money  Putts  Purse  Sandies  Score  \
117798        294.6  80.56  2070000.0  1.667    NaN    66.67  68.75   
17525         313.6  73.61   858667.0  1.639    NaN   100.00  69.00   
38616         308.0  70.83   858667.0  1.569    NaN   100.00  69.00   
127350        305.8  70.83   858667.0  1.597    NaN    57.14  69.00   
40810         294.8  65.28   403938.0  1.458    NaN    66.67  69.25   
54817         296.5  70.83   403938.0  1.556    NaN    33.33  69.25   
123063        316.3  66.67   403938.0  1.556    NaN    63.64  69.25   
126086        283.1  68.06   403938.0  1.556    NaN    50.00  69.25   
68306         308.4  70.83   310500.0  1.694    NaN    71.43  69.50   
94045         299.9  61.11   310500.0  1.472    NaN    71.43  69.50   

statistic  Scrambling  year  month  
117798          50.00  2019      4  
17525           68.42  2019      4  
38616           80.95  2019      4  
127350          57.14  2019      4  
40810           84.00  2019      4  
54817           71.43  2019      4  
123063          79.17  2019      4  
126086          78.26  2019      4  
68306           80.95  2019      4  
94045           67.86  2019      4

In [31]:
#Fairways
winners19_Fairways_model = sm.OLS.from_formula("Score ~ Fairways", data=df_2019_winners)
winners19_Fairways_results = winners19_Fairways_model.fit()
print("Score + Fairways | P-value:", winners19_Fairways_results.pvalues[1]/2, "| R-squared:", winners19_Fairways_results.rsquared)

#Distance
winners19_Distance_model = sm.OLS.from_formula("Score ~ Distance", data=df_2019_winners)
winners19_Distance_results = winners19_Distance_model.fit()
print("Score + Distance | P-value:", winners19_Distance_results.pvalues[1]/2, "| R-squared:", winners19_Distance_results.rsquared)

#GIR
winners19_GIR_model = sm.OLS.from_formula("Score ~ GIR", data=df_2019_winners)
winners19_GIR_results = winners19_GIR_model.fit()
print("Score + GIR | P-value:", winners19_GIR_results.pvalues[1]/2, "| R-squared:", winners19_GIR_results.rsquared)

#Putts
winners19_Putts_model = sm.OLS.from_formula("Score ~ Putts", data=df_2019_winners)
winners19_Putts_results = winners19_Putts_model.fit()
print("Score + Putts | P-value:", winners19_Putts_results.pvalues[1]/2, "| R-squared:", winners19_Putts_results.rsquared)

#Sandies
winners19_Sandies_model = sm.OLS.from_formula("Score ~ Sandies", data=df_2019_winners)
winners19_Sandies_results = winners19_Sandies_model.fit()
print("Score + Sandies | P-value:", winners19_Sandies_results.pvalues[1]/2, "| R-squared:", winners19_Sandies_results.rsquared)

#Scrambling
winners19_Scrambling_model = sm.OLS.from_formula("Score ~ Scrambling", data=df_2019_winners)
winners19_Scrambling_results = winners19_Scrambling_model.fit()
print("Score + Scrambling | P-value:", winners19_Scrambling_results.pvalues[1]/2, "| R-squared:", winners19_Scrambling_results.rsquared)

#Money
winners19_Money_model = sm.OLS.from_formula("Score ~ Money", data=df_2019_winners)
winners19_Money_results = winners19_Money_model.fit()
print("Score + Money | P-value:", winners19_Money_results.pvalues[1]/2, "| R-squared:", winners19_Money_results.rsquared)


#Note:
#P-value is the probability of getting a sample proportion at least this extreme.  Ranges from zero to 100%. 
#The P>|t| value in the OLS summary reflects a two-sided P-value, and needs to be divided by 2 to reflect the single sided value.
#P-values quoted in the summary of this notebook reflect the single sided P-value

Score + Fairways | P-value: 0.08540572714124003 | R-squared: 0.2205895551707081
Score + Distance | P-value: 0.47479371262309983 | R-squared: 0.000531671540754286
Score + GIR | P-value: 0.0035442625261141794 | R-squared: 0.616955220664926
Score + Putts | P-value: 0.14386293137365389 | R-squared: 0.13950084050055178
Score + Sandies | P-value: 0.24288836874753833 | R-squared: 0.06257362073341999
Score + Scrambling | P-value: 0.03227008456360929 | R-squared: 0.36455719703955947
Score + Money | P-value: 0.0004258759264315939 | R-squared: 0.769735230545112


### Summary

The results are different players who scored the lowest at the Masters over the last decade
- Nearly all the P-values are above alpha of 0.05 except for 2019 GIR (0.004) and 2019 Scrambling (0.032)
- The highest R-squared was 2019 GIR (0.617)
- Note that the summary analysis excludes Money, because the Money represents how much money they won at the Masters Tournament


|		|	2017 P-value	|	2017 R-squared	|	2018 P-value	|	2018 R-squared	|	2019 P-value	|	2019 R-squared	|
| :- | :-: | :-: | :-: | :-: | :-: | :-: |
|	Score + Fairways	|	0.156	|	0.127	|	0.197	|	0.092	|	0.085	|	0.221	|
|	Score + Distance	|	0.495	|	0.000	|	0.214	|	0.080	|	0.475	|	0.001	|
|	Score + GIR	|	0.133	|	0.151	|	0.358	|	0.017	|	0.004	|	0.617	|
|	Score + Putts	|	0.274	|	0.047	|	0.053	|	0.293	|	0.144	|	0.140	|
|	Score + Sandies	|	0.457	|	0.002	|	0.418	|	0.006	|	0.243	|	0.063	|
|	Score + Scrambling	|	0.401	|	0.008	|	0.400	|	0.008	|	0.032	|	0.365	|
|	Score + Money	|	0.000	|	0.811	|	0.000	|	0.798	|	0.000	|	0.770	|


## Summary of findings

So far the results are not ideal. Generally, the P-values are high and the correlations are low for most variables.  There isn't much consistency in variables between low scorers at prior Masters tournaments and top finishers over the last 3 years.

|		|	P-value	|	R-squared	|
| :- | :-: | :-: |
|	Top finishers at the 2019 Masters: Score + GIR	|	0.004	|	0.617	|
|	Top finishers at the 2019 Masters: Score + Scrambling	|	0.032	|	0.365	|
|	Lowest scorers from previous Masters: Score + Putts	|	0.036	|	0.201	|

## Reducing the data down to the last 2 years

In [32]:
df_2020winner = df[(df.year >= 2018) & (df.year <= 2020) & (df.Money >= 1)] #Also removing player's tournaments where they did not win any money
df_2020winner.head()

statistic     player_name        date                     tournament  \
230        Aaron Baddeley  2018-02-04  Waste Management Phoenix Open   
231        Aaron Baddeley  2018-02-11       AT&T Pebble Beach Pro-Am   
232        Aaron Baddeley  2018-02-18                   Genesis Open   
233        Aaron Baddeley  2018-03-11           Valspar Championship   
234        Aaron Baddeley  2018-04-01                   Houston Open   

statistic  Fairways  Distance    GIR     Money  Putts  Purse  Sandies  Score  \
230           48.21     289.4  62.50   51060.0  1.556   0.74    70.00  69.00   
231           61.82     288.3  63.89   16576.0  1.569   0.22    50.00  71.00   
232           46.43     307.1  48.61  133200.0  1.431   1.85    28.57  69.50   
233           40.38     301.0  56.94   15431.0  1.597   0.24    60.00  71.50   
234           67.86     290.3  69.44   13440.0  1.653   0.19    80.00  71.75   

statistic  Scrambling  year  month  
230             74.07  2018      2  
231             65.38  2018      2  
232             78.38  2018      2  
233             70.97  2018      3  
234             63.64  2018      4

## Aggregating the data by player

In [33]:
aggregation_functions = {'Score': 'mean', 'Fairways': 'mean', 'Distance': 'mean', 'GIR': 'mean', 'Putts': 'mean', 'Sandies': 'mean', 'Scrambling': 'mean', 'Money': 'sum'} #'player_name': 'first'
df_2020winner_agg = df.groupby(df_2020winner['player_name']).aggregate(aggregation_functions)
df_2020winner_agg.nsmallest(30, 'Score')

Score   Fairways    Distance        GIR  \
player_name                                                            
Andres Romero            67.000000  51.790000  302.500000  69.440000   
Jerry Kelly              67.000000  67.860000  288.000000  73.610000   
Jason Gore               67.250000  71.430000  282.800000  72.220000   
John Merrick             67.750000  82.140000  290.400000  79.170000   
Jason Bohn               68.000000  71.430000  259.500000  68.060000   
Y.E. Yang                68.000000  69.640000  310.100000  72.220000   
Jim Herman               68.250000  69.640000  292.400000  73.610000   
Brendon Todd             68.437500  72.767500  280.500000  65.970000   
Parker McLachlin         68.500000  66.517500  281.375000  72.222500   
Stuart Appleby           68.562500  73.660000  279.850000  62.845000   
Wes Roach                68.562500  65.402500  307.025000  68.923750   
Joey Garber              68.721667  67.858333  301.350000  72.765000   
David Hearn              68.727273  65.584545  286.918182  73.231818   
Brett Stegmaier          68.750000  57.144000  298.780000  73.334000   
Geoff Ogilvy             68.750000  85.710000  293.400000  75.000000   
José de Jesús Rodríguez  68.785714  58.927143  294.628571  66.667143   
Matthew Wolff            68.791667  59.521667  307.766667  65.276667   
Justin Thomas            68.805556  57.420556  307.427273  69.636667   
Rory McIlroy             68.830645  58.963548  316.876667  67.876129   
Adam Svensson            68.875000  65.714000  296.950000  72.222000   
Webb Simpson             68.888158  65.531053  291.871053  67.251053   
Dustin Johnson           68.925676  57.284054  311.300000  69.032162   
Bronson Burgoon          68.986842  67.550526  302.883333  72.076316   
Sungjae Im               69.000000  65.648800  299.595652  69.222800   
Jon Rahm                 69.015625  62.385000  307.435484  69.921250   
Denny McCarthy           69.034091  68.447273  292.554545  69.760455   
Brooks Koepka            69.040323  60.583226  309.282143  68.772258   
Zack Sucher              69.041667  62.795000  309.916667  68.055000   
Sebastián Muñoz          69.096154  60.578462  303.076923  70.299231   
Anders Albertson         69.111667  62.798333  299.100000  73.610000   

                            Putts     Sandies  Scrambling       Money  
player_name                                                            
Andres Romero            1.444000   44.440000   72.730000    337560.0  
Jerry Kelly              1.625000   66.670000   78.950000    108500.0  
Jason Gore               1.556000   50.000000   65.000000     92960.0  
John Merrick             1.583000    0.000000   73.330000     97772.0  
Jason Bohn               1.514000  100.000000   69.570000     73660.0  
Y.E. Yang                1.625000   42.860000   70.000000     31040.0  
Jim Herman               1.639000   37.500000   68.420000     19488.0  
Brendon Todd             1.496750   79.807500   74.737500    255132.0  
Parker McLachlin         1.590000   60.120000   62.217500    233341.0  
Stuart Appleby           1.527750   79.615000   76.217500    170244.0  
Wes Roach                1.578375   48.021250   65.777500    930807.0  
Joey Garber              1.601167   44.445000   64.250000    355681.0  
David Hearn              1.603273   50.714545   59.709091    670543.0  
Brett Stegmaier          1.600000   71.906000   60.950000    211928.0  
Geoff Ogilvy             1.653000   33.330000   72.220000     39116.0  
José de Jesús Rodríguez  1.535714   57.994286   66.770000    293997.0  
Matthew Wolff            1.576500   32.500000   61.596667    352545.0  
Justin Thomas            1.576778   56.131667   63.268611  11811383.0  
Rory McIlroy             1.561387   53.248710   65.110645  14275016.0  
Adam Svensson            1.594600   41.787000   63.417000    451280.0  
Webb Simpson             1.561789   65.709737   67.551316  10899228.0  
Dustin Johnson           1.579703   62.522703   62.829459  14831104.

In [34]:
df_2020winner_agg.head()

Score   Fairways    Distance        GIR     Putts  \
player_name                                                             
Aaron Baddeley  70.214286  55.461905  293.066667  64.020476  1.570857   
Aaron Wise      69.651786  64.709286  303.189286  70.138929  1.619536   
Abraham Ancer   69.750000  67.656486  296.311765  67.941622  1.594919   
Adam Hadwin     69.986486  66.744865  290.232353  68.131081  1.596811   
Adam Long       69.687500  69.343333  294.708333  63.889167  1.552083   

                  Sandies  Scrambling      Money  
player_name                                       
Aaron Baddeley  57.583810   64.538571  1153971.0  
Aaron Wise      46.785714   54.126071  4321645.0  
Abraham Ancer   50.063514   63.862703  4277384.0  
Adam Hadwin     54.415946   61.842162  3961044.0  
Adam Long       56.904167   63.207500  1742343.0

## Comparing independent variables

#### Comparing the variables to "Score" (average number of strokes per completed round)

In [35]:
#Fairways
Score_Fairways_model = sm.OLS.from_formula("Score ~ Fairways", data=df_2020winner_agg)
Score_Fairways_results = Score_Fairways_model.fit()
print("Score + Fairways | P-value:", Score_Fairways_results.pvalues[1]/2, "| R-squared:", Score_Fairways_results.rsquared)

#Distance
Score_Distance_model = sm.OLS.from_formula("Score ~ Distance", data=df_2020winner_agg)
Score_Distance_results = Score_Distance_model.fit()
print("Score + Distance | P-value:", Score_Distance_results.pvalues[1]/2, "| R-squared:", Score_Distance_results.rsquared)

#GIR
Score_GIR_model = sm.OLS.from_formula("Score ~ GIR", data=df_2020winner_agg)
Score_GIR_results = Score_GIR_model.fit()
print("Score + GIR | P-value:", Score_GIR_results.pvalues[1]/2, "| R-squared:", Score_GIR_results.rsquared)

#Putts
Score_Putts_model = sm.OLS.from_formula("Score ~ Putts", data=df_2020winner_agg)
Score_Putts_results = Score_Putts_model.fit()
print("Score + Putts | P-value:", Score_Putts_results.pvalues[1]/2, "| R-squared:", Score_Putts_results.rsquared)

#Sandies
Score_Sandies_model = sm.OLS.from_formula("Score ~ Sandies", data=df_2020winner_agg)
Score_Sandies_results = Score_Sandies_model.fit()
print("Score + Sandies | P-value:", Score_Sandies_results.pvalues[1]/2, "| R-squared:", Score_Sandies_results.rsquared)

#Scrambling
Score_Scrambling_model = sm.OLS.from_formula("Score ~ Scrambling", data=df_2020winner_agg)
Score_Scrambling_results = Score_Scrambling_model.fit()
print("Score + Scrambling | P-value:", Score_Scrambling_results.pvalues[1]/2, "| R-squared:", Score_Scrambling_results.rsquared)

#Money
Score_Money_model = sm.OLS.from_formula("Score ~ Money", data=df_2020winner_agg)
Score_Money_results = Score_Money_model.fit()
print("Score + Money | P-value:", Score_Money_results.pvalues[1]/2, "| R-squared:", Score_Money_results.rsquared)


#Note:
#P-value is the probability of getting a sample proportion at least this extreme.  Ranges from zero to 100%. 
#The P>|t| value in the OLS summary reflects a two-sided P-value, and needs to be divided by 2 to reflect the single sided value.
#P-values quoted in the summary of this notebook reflect the single sided P-value

Score + Fairways | P-value: 0.0046821950855813425 | R-squared: 0.022896560833514434
Score + Distance | P-value: 0.001852749598786064 | R-squared: 0.028577606637016806
Score + GIR | P-value: 1.9101645065828996e-27 | R-squared: 0.3292806086437271
Score + Putts | P-value: 1.2441167889370516e-09 | R-squared: 0.11480449519486935
Score + Sandies | P-value: 0.041579068038456554 | R-squared: 0.010245821008425526
Score + Scrambling | P-value: 8.401530269278429e-29 | R-squared: 0.34338699649251914
Score + Money | P-value: 8.47088898461933e-06 | R-squared: 0.06150984862995712


In [36]:
#Fairways
Money_Fairways_model = sm.OLS.from_formula("Money ~ Fairways", data=df_2020winner_agg)
Money_Fairways_results = Money_Fairways_model.fit()
print("Money + Fairways | P-value:", Money_Fairways_results.pvalues[1]/2, "| R-squared:", Money_Fairways_results.rsquared)

#Distance
Money_Distance_model = sm.OLS.from_formula("Money ~ Distance", data=df_2020winner_agg)
Money_Distance_results = Money_Distance_model.fit()
print("Money + Distance | P-value:", Money_Distance_results.pvalues[1]/2, "| R-squared:", Money_Distance_results.rsquared)

#GIR
Money_GIR_model = sm.OLS.from_formula("Money ~ GIR", data=df_2020winner_agg)
Money_GIR_results = Money_GIR_model.fit()
print("Money + GIR | P-value:", Money_GIR_results.pvalues[1]/2, "| R-squared:", Money_GIR_results.rsquared)

#Putts
Money_Putts_model = sm.OLS.from_formula("Money ~ Putts", data=df_2020winner_agg)
Money_Putts_results = Money_Putts_model.fit()
print("Money + Putts | P-value:", Money_Putts_results.pvalues[1]/2, "| R-squared:", Money_Putts_results.rsquared)

#Sandies
Money_Sandies_model = sm.OLS.from_formula("Money ~ Sandies", data=df_2020winner_agg)
Money_Sandies_results = Money_Sandies_model.fit()
print("Money + Sandies | P-value:", Money_Sandies_results.pvalues[1]/2, "| R-squared:", Money_Sandies_results.rsquared)

#Scrambling
Money_Scrambling_model = sm.OLS.from_formula("Money ~ Scrambling", data=df_2020winner_agg)
Money_Scrambling_results = Money_Scrambling_model.fit()
print("Money + Scrambling | P-value:", Money_Scrambling_results.pvalues[1]/2, "| R-squared:", Money_Scrambling_results.rsquared)

#Money
Money_Score_model = sm.OLS.from_formula("Money ~ Score", data=df_2020winner_agg)
Money_Score_results = Money_Score_model.fit()
print("Money + Score | P-value:", Money_Score_results.pvalues[1]/2, "| R-squared:", Money_Score_results.rsquared)


Money + Fairways | P-value: 0.1894646566085978 | R-squared: 0.002652295179736641
Money + Distance | P-value: 1.1417847103729464e-09 | R-squared: 0.1156851158568385
Money + GIR | P-value: 0.0058611230597433615 | R-squared: 0.02155628692597822
Money + Putts | P-value: 0.07946860796274853 | R-squared: 0.006784101132768172
Money + Sandies | P-value: 0.0253325205374173 | R-squared: 0.013016766319043027
Money + Scrambling | P-value: 0.0034739569925010694 | R-squared: 0.02468623478122689
Money + Score | P-value: 8.470888984614513e-06 | R-squared: 0.06150984862995723


In [37]:
df_2020winner_agg.corr()

Score  Fairways  Distance       GIR     Putts   Sandies  \
Score       1.000000 -0.151316 -0.169049 -0.573830  0.338828 -0.101222   
Fairways   -0.151316  1.000000 -0.440184  0.316600  0.245902 -0.234139   
Distance   -0.169049 -0.440184  1.000000  0.208092  0.135373 -0.086258   
GIR        -0.573830  0.316600  0.208092  1.000000  0.453204 -0.087838   
Putts       0.338828  0.245902  0.135373  0.453204  1.000000 -0.197015   
Sandies    -0.101222 -0.234139 -0.086258 -0.087838 -0.197015  1.000000   
Scrambling -0.585992 -0.019394 -0.017175  0.040143 -0.546536  0.351386   
Money      -0.248012 -0.051500  0.340125  0.146821 -0.082366  0.114091   

            Scrambling     Money  
Score        -0.585992 -0.248012  
Fairways     -0.019394 -0.051500  
Distance     -0.017175  0.340125  
GIR           0.040143  0.146821  
Putts        -0.546536 -0.082366  
Sandies       0.351386  0.114091  
Scrambling    1.000000  0.157119  
Money         0.157119  1.000000

## Creating the model using Score as the dependent variable

In [38]:
Model = sm.OLS.from_formula("Score ~ GIR + Scrambling + Putts", data=df_2020winner_agg)
Results = Model.fit()
print(Results.summary())

                            OLS Regression Results                            
Dep. Variable:                  Score   R-squared:                       0.812
Model:                            OLS   Adj. R-squared:                  0.810
Method:                 Least Squares   F-statistic:                     418.2
Date:                Sun, 11 Oct 2020   Prob (F-statistic):          5.75e-105
Time:                        11:20:19   Log-Likelihood:                -146.44
No. Observations:                 294   AIC:                             300.9
Df Residuals:                     290   BIC:                             315.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     64.4367      1.369     47.061      0.0

In [39]:
df_2020winner_agg['Prediction-Score'] = 64.4367 + (-0.1984*df_2020winner_agg.GIR) + (-0.0408*df_2020winner_agg.Scrambling) + (13.3478*df_2020winner_agg.Putts)
df_2020winner_agg.nsmallest(20, 'Prediction-Score') #use .nsmallest for lowest score

Score   Fairways    Distance        GIR  \
player_name                                                            
John Merrick             67.750000  82.140000  290.400000  79.170000   
Andres Romero            67.000000  51.790000  302.500000  69.440000   
Jason Gore               67.250000  71.430000  282.800000  72.220000   
Brendon Todd             68.437500  72.767500  280.500000  65.970000   
Jerry Kelly              67.000000  67.860000  288.000000  73.610000   
Jason Bohn               68.000000  71.430000  259.500000  68.060000   
Anders Albertson         69.111667  62.798333  299.100000  73.610000   
Geoff Ogilvy             68.750000  85.710000  293.400000  75.000000   
Joey Garber              68.721667  67.858333  301.350000  72.765000   
Brett Stegmaier          68.750000  57.144000  298.780000  73.334000   
Tom Lehman               69.250000  87.500000  280.300000  77.780000   
Parker McLachlin         68.500000  66.517500  281.375000  72.222500   
Adam Svensson            68.875000  65.714000  296.950000  72.222000   
David Hearn              68.727273  65.584545  286.918182  73.231818   
Jim Herman               68.250000  69.640000  292.400000  73.610000   
Y.E. Yang                68.000000  69.640000  310.100000  72.220000   
José de Jesús Rodríguez  68.785714  58.927143  294.628571  66.667143   
Chad Collins             69.875000  60.715000  277.800000  77.085000   
Sungjae Im               69.000000  65.648800  299.595652  69.222800   
Justin Thomas            68.805556  57.420556  307.427273  69.636667   

                            Putts     Sandies  Scrambling       Money  \
player_name                                                             
John Merrick             1.583000    0.000000   73.330000     97772.0   
Andres Romero            1.444000   44.440000   72.730000    337560.0   
Jason Gore               1.556000   50.000000   65.000000     92960.0   
Brendon Todd             1.496750   79.807500   74.737500    255132.0   
Jerry Kelly              1.625000   66.670000   78.950000    108500.0   
Jason Bohn               1.514000  100.000000   69.570000     73660.0   
Anders Albertson         1.622500   75.000000   70.568333    209712.0   
Geoff Ogilvy             1.653000   33.330000   72.220000     39116.0   
Joey Garber              1.601167   44.445000   64.250000    355681.0   
Brett Stegmaier          1.600000   71.906000   60.950000    211928.0   
Tom Lehman               1.653000   33.330000   56.250000     13952.0   
Parker McLachlin         1.590000   60.120000   62.217500    233341.0   
Adam Svensson            1.594600   41.787000   63.417000    451280.0   
David Hearn              1.603273   50.714545   59.709091    670543.0   
Jim Herman               1.639000   37.500000   68.420000     19488.0   
Y.E. Yang                1.625000   42.860000   70.000000     31040.0   
José de Jesús Rodríguez  1.535714   57.994286   66.770000    293997.0   
Chad Collins             1.645500   41.665000   51.115000     28105.0   
Sungjae Im               1.575080   53.309200   66.281600   3098461.0   
Justin Thomas            1.576778   56.131667   63.268611  11811383.0   

                         Prediction-Score  
player_name                                
John Merrick                    66.867075  
Andres Romero                   66.966643  
Jason Gore                      68.225429  
Brendon Todd                    68.277282  
Jerry Kelly                     68.301491  
Jason Bohn                      68.303709  
Anders Albertson                68.610094  
Geoff Ogilvy                    68.674037  
Joey Garber                     68.750776  
Brett Stegmaier                 68.756954  
Tom Lehman                      68.774061  
Parker McLachlin                68.792284  
Adam Svensson                   68.804843  
David Hearn                     68.871540  
Jim Herman                      68.917984  
Y.E. Yang                       68.942427  
José de Jesús Rodríguez         68.984130  
C

## Creating the model using Money as the dependent variable

In [40]:
#Model = sm.OLS.from_formula("Money ~ GIR + Scrambling + Putts", data=df_2020winner)
#Results = Model.fit()
#print(Results.summary())

In [41]:
#df_2020winner['Prediction-Money'] = 946500 + (10020*df_2020winner.GIR) + (1727.52*df_2020winner.Scrambling) + (-1012000*df_2020winner.Putts)
#df_2020winner.nlargest(10, 'Prediction-Money') #use .nlargest for largest amount of Money to be won

In [42]:
Model = sm.OLS.from_formula("Money ~ GIR + Scrambling + Putts", data=df_2020winner_agg)
Results = Model.fit()
print(Results.summary())

                            OLS Regression Results                            
Dep. Variable:                  Money   R-squared:                       0.053
Model:                            OLS   Adj. R-squared:                  0.043
Method:                 Least Squares   F-statistic:                     5.401
Date:                Sun, 11 Oct 2020   Prob (F-statistic):            0.00125
Time:                        11:20:20   Log-Likelihood:                -4768.8
No. Observations:                 294   AIC:                             9546.
Df Residuals:                     290   BIC:                             9560.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   4.237e+06   9.22e+06      0.460      0.6

In [43]:
df_2020winner_agg['Prediction-Money'] = 4.237e+06 + (1.459e+05*df_2020winner_agg.GIR) + (4.044e+04*df_2020winner_agg.Scrambling) + (-9.039e+06*df_2020winner_agg.Putts)
df_2020winner_agg.nlargest(20, 'Prediction-Money') #use .nlargest for largest amount of Money to be won

Score   Fairways    Distance        GIR  \
player_name                                                            
John Merrick             67.750000  82.140000  290.400000  79.170000   
Andres Romero            67.000000  51.790000  302.500000  69.440000   
Jerry Kelly              67.000000  67.860000  288.000000  73.610000   
Brendon Todd             68.437500  72.767500  280.500000  65.970000   
Jason Gore               67.250000  71.430000  282.800000  72.220000   
Jason Bohn               68.000000  71.430000  259.500000  68.060000   
Anders Albertson         69.111667  62.798333  299.100000  73.610000   
Geoff Ogilvy             68.750000  85.710000  293.400000  75.000000   
Joey Garber              68.721667  67.858333  301.350000  72.765000   
Brett Stegmaier          68.750000  57.144000  298.780000  73.334000   
Jim Herman               68.250000  69.640000  292.400000  73.610000   
Adam Svensson            68.875000  65.714000  296.950000  72.222000   
Tom Lehman               69.250000  87.500000  280.300000  77.780000   
Parker McLachlin         68.500000  66.517500  281.375000  72.222500   
Y.E. Yang                68.000000  69.640000  310.100000  72.220000   
David Hearn              68.727273  65.584545  286.918182  73.231818   
José de Jesús Rodríguez  68.785714  58.927143  294.628571  66.667143   
Sungjae Im               69.000000  65.648800  299.595652  69.222800   
Justin Thomas            68.805556  57.420556  307.427273  69.636667   
Rob Oppenheim            69.708333  59.713333  287.400000  65.278333   

                            Putts     Sandies  Scrambling       Money  \
player_name                                                             
John Merrick             1.583000    0.000000   73.330000     97772.0   
Andres Romero            1.444000   44.440000   72.730000    337560.0   
Jerry Kelly              1.625000   66.670000   78.950000    108500.0   
Brendon Todd             1.496750   79.807500   74.737500    255132.0   
Jason Gore               1.556000   50.000000   65.000000     92960.0   
Jason Bohn               1.514000  100.000000   69.570000     73660.0   
Anders Albertson         1.622500   75.000000   70.568333    209712.0   
Geoff Ogilvy             1.653000   33.330000   72.220000     39116.0   
Joey Garber              1.601167   44.445000   64.250000    355681.0   
Brett Stegmaier          1.600000   71.906000   60.950000    211928.0   
Jim Herman               1.639000   37.500000   68.420000     19488.0   
Adam Svensson            1.594600   41.787000   63.417000    451280.0   
Tom Lehman               1.653000   33.330000   56.250000     13952.0   
Parker McLachlin         1.590000   60.120000   62.217500    233341.0   
Y.E. Yang                1.625000   42.860000   70.000000     31040.0   
David Hearn              1.603273   50.714545   59.709091    670543.0   
José de Jesús Rodríguez  1.535714   57.994286   66.770000    293997.0   
Sungjae Im               1.575080   53.309200   66.281600   3098461.0   
Justin Thomas            1.576778   56.131667   63.268611  11811383.0   
Rob Oppenheim            1.543833   59.030000   71.508333    128418.0   

                         Prediction-Score  Prediction-Money  
player_name                                                  
John Merrick                    66.867075      4.444631e+06  
Andres Romero                   66.966643      4.257181e+06  
Jerry Kelly                     68.301491      3.481062e+06  
Brendon Todd                    68.277282      3.355284e+06  
Jason Gore                      68.225429      3.337814e+06  
Jason Bohn                      68.303709      3.295319e+06  
Anders Albertson                68.610094      3.164705e+06  
Geoff Ogilvy                    68.674037      3.158610e+06  
Joey Garber                     68.750776      2.978738e+06  
Brett Stegmaier                 68.756954      2.938849e+06  
Jim Herman                      68.917984      2.928683e+06  
Adam Svensson                  

## Creating the model only using previous Masters data

### Reducing the data

In [44]:
df_2020winner_mastersonly = df[(df.year >= 2018) & (df.year <= 2020) & (df.Money >= 1) & (df_2020winner['tournament'] == 'Masters Tournament')]
df_2020winner_mastersonly.head()

statistic  player_name        date          tournament  Fairways  Distance  \
315         Aaron Wise  2019-04-14  Masters Tournament     66.07     305.8   
455        Adam Hadwin  2018-04-08  Masters Tournament     69.64     291.9   
760         Adam Scott  2018-04-08  Masters Tournament     69.64     290.3   
778         Adam Scott  2019-04-14  Masters Tournament     71.43     310.3   
1233        Alex Noren  2019-04-14  Masters Tournament     76.79     291.0   

statistic    GIR     Money  Putts  Purse  Sandies  Score  Scrambling  year  \
315        68.06  184000.0  1.597    NaN    50.00  70.25       56.52  2019   
455        69.44   93775.0  1.667   0.85    83.33  71.75       54.55  2018   
760        68.06   63663.0  1.736   0.58    80.00  72.25       56.52  2018   
778        75.00  161000.0  1.667    NaN    33.33  70.50       66.67  2019   
1233       55.56   25415.0  1.639    NaN    33.33  74.00       56.25  2019   

statistic  month  
315            4  
455            4  
760            4  
778            4  
1233           4

In [45]:
aggregation_functions = {'Score': 'mean', 'Fairways': 'mean', 'Distance': 'mean', 'GIR': 'mean', 'Putts': 'mean', 'Sandies': 'mean', 'Scrambling': 'mean', 'Money': 'sum'} #'player_name': 'first'
df_2020winner_mastersonly_agg = df.groupby(df_2020winner_mastersonly['player_name']).aggregate(aggregation_functions)
df_2020winner_mastersonly_agg.nsmallest(20, 'Score')

Score  Fairways  Distance     GIR   Putts  Sandies  \
player_name                                                               
Brooks Koepka       69.000    69.640    313.60  73.610  1.6390  100.000   
Rickie Fowler       69.000    74.110    295.55  67.360  1.5205   41.665   
Jon Rahm            69.375    75.000    301.90  69.445  1.6250   73.215   
Patrick Cantlay     69.500    64.290    299.90  61.110  1.4720   71.430   
Dustin Johnson      69.625    63.390    306.45  69.445  1.6180   75.000   
Jordan Spieth       69.750    66.965    289.85  70.830  1.6180   76.190   
Tony Finau          69.750    61.610    303.30  63.195  1.5490   41.820   
Bubba Watson        69.875    70.535    298.45  73.610  1.6805   70.000   
Patrick Reed        69.875    75.000    297.70  62.500  1.5275   50.000   
Aaron Wise          70.250    66.070    305.80  68.060  1.5970   50.000   
Rory McIlroy        70.250    60.715    308.50  61.110  1.5695   52.500   
Francesco Molinari  70.375    75.000    291.95  67.360  1.5900   66.670   
Jason Day           70.375    65.180    294.30  68.055  1.6115   58.330   
Webb Simpson        70.375    75.000    283.05  63.195  1.5560   52.780   
Justin Rose         70.500    73.210    294.40  73.610  1.6810   60.000   
Justin Thomas       70.500    66.960    307.30  75.000  1.6740   55.000   
Patton Kizzire      70.500    71.430    299.90  66.670  1.5690   75.000   
Tiger Woods         70.500    58.035    294.20  73.615  1.6460   50.000   
Henrik Stenson      70.625    79.465    287.40  66.665  1.6180   41.665   
Charley Hoffman     70.750    66.965    292.25  69.445  1.6250   57.500   

                    Scrambling      Money  
player_name                                
Brooks Koepka           68.420   858667.0  
Rickie Fowler           72.710  1498500.0  
Jon Rahm                70.910   838500.0  
Patrick Cantlay         67.860   310500.0  
Dustin Johnson          73.085  1144667.0  
Jordan Spieth           66.820   855956.0  
Tony Finau              70.620   689938.0  
Bubba Watson            65.340   611775.0  
Patrick Reed            62.915  2035488.0  
Aaron Wise              56.520   184000.0  
Rory McIlroy            58.300   494331.0  
Francesco Molinari      71.545   532088.0  
Jason Day               69.715   532088.0  
Webb Simpson            64.130   532088.0  
Justin Rose             52.630   231000.0  
Justin Thomas           64.240   395900.0  
Patton Kizzire          62.500   161000.0  
Tiger Woods             50.000  2133663.0  
Henrik Stenson          58.200   441863.0  
Charley Hoffman         60.000   309200.0

In [46]:
df_2020winner_mastersonly_agg.head()

Score  Fairways  Distance    GIR   Putts  Sandies  Scrambling  \
player_name                                                                     
Aaron Wise     70.250    66.070     305.8  68.06  1.5970   50.000      56.520   
Adam Hadwin    71.750    69.640     291.9  69.44  1.6670   83.330      54.550   
Adam Scott     71.375    70.535     300.3  71.53  1.7015   56.665      61.595   
Alex Noren     74.000    76.790     291.0  55.56  1.6390   33.330      56.250   
Andrew Landry  72.000    80.360     287.6  65.28  1.5830   66.670      60.000   

                  Money  
player_name              
Aaron Wise     184000.0  
Adam Hadwin     93775.0  
Adam Scott     224663.0  
Alex Noren      25415.0  
Andrew Landry   37950.0

### Using Score as the dependent variable

In [47]:
Model = sm.OLS.from_formula("Score ~ GIR + Scrambling + Putts", data=df_2020winner_mastersonly_agg)
Results = Model.fit()
print(Results.summary())

                            OLS Regression Results                            
Dep. Variable:                  Score   R-squared:                       0.827
Model:                            OLS   Adj. R-squared:                  0.819
Method:                 Least Squares   F-statistic:                     98.98
Date:                Sun, 11 Oct 2020   Prob (F-statistic):           1.32e-23
Time:                        11:20:20   Log-Likelihood:                -52.837
No. Observations:                  66   AIC:                             113.7
Df Residuals:                      62   BIC:                             122.4
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     59.0038      2.348     25.126      0.0

In [48]:
df_2020winner_mastersonly_agg['Prediction-Score-MastersOnly'] = 59.0038 + (-0.2305*df_2020winner_mastersonly_agg.GIR) + (-0.0149*df_2020winner_mastersonly_agg.Scrambling) + (17.4504*df_2020winner_mastersonly_agg.Putts)
df_2020winner_mastersonly_agg.nsmallest(30, 'Prediction-Score-MastersOnly') #use .nsmallest for lowest score

Score  Fairways  Distance     GIR   Putts  Sandies  \
player_name                                                               
Rickie Fowler       69.000    74.110    295.55  67.360  1.5205   41.665   
Patrick Cantlay     69.500    64.290    299.90  61.110  1.4720   71.430   
Brooks Koepka       69.000    69.640    313.60  73.610  1.6390  100.000   
Jordan Spieth       69.750    66.965    289.85  70.830  1.6180   76.190   
Justin Thomas       70.500    66.960    307.30  75.000  1.6740   55.000   
Tiger Woods         70.500    58.035    294.20  73.615  1.6460   50.000   
Patton Kizzire      70.500    71.430    299.90  66.670  1.5690   75.000   
Dustin Johnson      69.625    63.390    306.45  69.445  1.6180   75.000   
Francesco Molinari  70.375    75.000    291.95  67.360  1.5900   66.670   
Jon Rahm            69.375    75.000    301.90  69.445  1.6250   73.215   
Patrick Reed        69.875    75.000    297.70  62.500  1.5275   50.000   
Aaron Wise          70.250    66.070    305.80  68.060  1.5970   50.000   
Bubba Watson        69.875    70.535    298.45  73.610  1.6805   70.000   
Jason Day           70.375    65.180    294.30  68.055  1.6115   58.330   
Tony Finau          69.750    61.610    303.30  63.195  1.5490   41.820   
Charley Hoffman     70.750    66.965    292.25  69.445  1.6250   57.500   
Louis Oosthuizen    70.750    66.965    295.20  68.055  1.6250   58.335   
Justin Rose         70.500    73.210    294.40  73.610  1.6810   60.000   
Webb Simpson        70.375    75.000    283.05  63.195  1.5560   52.780   
Andrew Landry       72.000    80.360    287.60  65.280  1.5830   66.670   
Kevin Kisner        71.375    74.105    286.30  63.885  1.5625   32.500   
Paul Casey          70.750    50.000    281.80  62.500  1.5560   14.290   
Matt Kuchar         71.000    72.320    288.70  65.275  1.5900   50.000   
Phil Mickelson      71.500    50.895    291.35  63.195  1.5695   75.000   
Henrik Stenson      70.625    79.465    287.40  66.665  1.6180   41.665   
Marc Leishman       71.125    57.140    296.35  63.890  1.5830   45.835   
Cameron Smith       71.125    56.250    295.45  62.500  1.5760   70.000   
Ian Poulter         71.500    73.215    292.40  69.445  1.6595   37.500   
Russell Henley      70.750    76.790    286.30  70.830  1.6670  100.000   
Adam Hadwin         71.750    69.640    291.90  69.440  1.6670   83.330   

                    Scrambling      Money  Prediction-Score-MastersOnly  
player_name                                                              
Rickie Fowler           72.710  1498500.0                     68.927274  
Patrick Cantlay         67.860   310500.0                     69.593820  
Brooks Koepka           68.420   858667.0                     69.618443  
Jordan Spieth           66.820   855956.0                     69.916614  
Justin Thomas           64.240   395900.0                     69.971094  
Tiger Woods             50.000  2133663.0                     70.013901  
Patton Kizzire          62.500   161000.0                     70.084793  
Dustin Johnson          73.085  1144667.0                     70.142508  
Francesco Molinari      71.545   532088.0                     70.157435  
Jon Rahm                70.910   838500.0                     70.297068  
Patrick Reed            62.915  2035488.0                     70.315602  
Aaron Wise              56.520   184000.0                     70.342111  
Bubba Watson            65.340   611775.0                     70.388526  
Jason Day               69.715   532088.0                     70.399689  
Tony Finau              70.620   689938.0                     70.415784  
Charley Hoffman         60.000   309200.0                     70.459627  
Louis Oosthuizen        76.515   309200.0                     70.533949  
Justin Rose             52.630   231000.0                     70.586630  
Webb Simpson            64.130   532088.0                     70.634638  
Andrew Landry           60.000    37950.0                     70.6

### Using Money as the dependent variable

In [49]:
Model = sm.OLS.from_formula("Money ~ GIR + Scrambling + Putts", data=df_2020winner_mastersonly_agg)
Results = Model.fit()
print(Results.summary())

                            OLS Regression Results                            
Dep. Variable:                  Money   R-squared:                       0.371
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                     12.21
Date:                Sun, 11 Oct 2020   Prob (F-statistic):           2.24e-06
Time:                        11:20:20   Log-Likelihood:                -934.56
No. Observations:                  66   AIC:                             1877.
Df Residuals:                      62   BIC:                             1886.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   2.666e+06   1.49e+06      1.792      0.0

In [50]:
df_2020winner_mastersonly_agg['Prediction-Money-MastersOnly'] = 2.666e+06 + (4.859e+04*df_2020winner_mastersonly_agg.GIR) + (6042.6003*df_2020winner_mastersonly_agg.Scrambling) + (-3.605e+06*df_2020winner_mastersonly_agg.Putts)
df_2020winner_mastersonly_agg.nlargest(20, 'Prediction-Money-MastersOnly') #use .nlargest for largest amount of Money to be won

Score  Fairways  Distance     GIR   Putts  Sandies  \
player_name                                                               
Rickie Fowler       69.000    74.110    295.55  67.360  1.5205   41.665   
Brooks Koepka       69.000    69.640    313.60  73.610  1.6390  100.000   
Patrick Cantlay     69.500    64.290    299.90  61.110  1.4720   71.430   
Jordan Spieth       69.750    66.965    289.85  70.830  1.6180   76.190   
Justin Thomas       70.500    66.960    307.30  75.000  1.6740   55.000   
Dustin Johnson      69.625    63.390    306.45  69.445  1.6180   75.000   
Francesco Molinari  70.375    75.000    291.95  67.360  1.5900   66.670   
Patton Kizzire      70.500    71.430    299.90  66.670  1.5690   75.000   
Tiger Woods         70.500    58.035    294.20  73.615  1.6460   50.000   
Jon Rahm            69.375    75.000    301.90  69.445  1.6250   73.215   
Jason Day           70.375    65.180    294.30  68.055  1.6115   58.330   
Bubba Watson        69.875    70.535    298.45  73.610  1.6805   70.000   
Tony Finau          69.750    61.610    303.30  63.195  1.5490   41.820   
Louis Oosthuizen    70.750    66.965    295.20  68.055  1.6250   58.335   
Patrick Reed        69.875    75.000    297.70  62.500  1.5275   50.000   
Aaron Wise          70.250    66.070    305.80  68.060  1.5970   50.000   
Charley Hoffman     70.750    66.965    292.25  69.445  1.6250   57.500   
Webb Simpson        70.375    75.000    283.05  63.195  1.5560   52.780   
Justin Rose         70.500    73.210    294.40  73.610  1.6810   60.000   
Andrew Landry       72.000    80.360    287.60  65.280  1.5830   66.670   

                    Scrambling      Money  Prediction-Score-MastersOnly  \
player_name                                                               
Rickie Fowler           72.710  1498500.0                     68.927274   
Brooks Koepka           68.420   858667.0                     69.618443   
Patrick Cantlay         67.860   310500.0                     69.593820   
Jordan Spieth           66.820   855956.0                     69.916614   
Justin Thomas           64.240   395900.0                     69.971094   
Dustin Johnson          73.085  1144667.0                     70.142508   
Francesco Molinari      71.545   532088.0                     70.157435   
Patton Kizzire          62.500   161000.0                     70.084793   
Tiger Woods             50.000  2133663.0                     70.013901   
Jon Rahm                70.910   838500.0                     70.297068   
Jason Day               69.715   532088.0                     70.399689   
Bubba Watson            65.340   611775.0                     70.388526   
Tony Finau              70.620   689938.0                     70.415784   
Louis Oosthuizen        76.515   309200.0                     70.533949   
Patrick Reed            62.915  2035488.0                     70.315602   
Aaron Wise              56.520   184000.0                     70.342111   
Charley Hoffman         60.000   309200.0                     70.459627   
Webb Simpson            64.130   532088.0                     70.634638   
Justin Rose             52.630   231000.0                     70.586630   
Andrew Landry           60.000    37950.0                     70.686743   

                    Prediction-Money-MastersOnly  
player_name                                       
Rickie Fowler                      896977.367813  
Brooks Koepka                      747549.612526  
Patrick Cantlay                    738825.756358  
Jordan Spieth                      678506.252046  
Justin Thomas                      663656.643272  
Dustin Johnson                     649065.992925  
Francesco Molinari                 639390.238464  
Patton Kizzire                     626912.818750  
Tiger Woods                        611252.865000  
Jon Rahm                           610688.337273  
Jason Day                          584594.829915  
Bubba Watson                       579330.903602  
Tony 